src/Microsoft.AspNetCore.Server.Kestrel/Http/Frame.cs

```diff
@@ -752,35 +779,66 @@ protected bool TakeStartLine(SocketInput input)
                 if (chFound == '?')
                 {
                     begin = scan;
-                    if (scan.Seek(ref _vectorSpaces) != ' ')
+                    if (scan.Seek(ref _vectorSpaces) == -1)
                     {
-                        return false;
+                        return RequestLineStatus.TargetIncomplete;
                     }
                     queryString = begin.GetAsciiString(scan);
                 }
 
+                if (pathBegin.GetLength(scan) == 0)
+                {
+                    ReportBadHttpRequest("Missing request target.");
+                    return RequestLineStatus.Done;
+                }
+
                 scan.Take();
                 begin = scan;
+                chFound = scan.Seek(ref _vectorCRs);
+                var scanned = begin.GetLength(scan);
+                if (chFound == -1 && scanned < _httpVersionLength)
+                {
+                    return RequestLineStatus.VersionIncomplete;
+                }
+                else if (chFound == '\r' && scanned == 0)
+                {
+                    ReportBadHttpRequest("Missing HTTP version.");
+                    return RequestLineStatus.Done;
+                }
+                else if ((chFound == '\r' &&  scanned < _httpVersionLength) || scanned > _httpVersionLength)
```

Comment:
`else if (scanned < _httpVersionLength || scanned > _httpVersionLength)` is easier to read.

You've already returned `RequestLineStatus.VersionIncomplete` `if (chFound == -1 && scanned < _httpVersionLength)`.